src/performance_warnings.jl

```diff
@@ -0,0 +1,44 @@
+
+const PERFORMANCE_WARNINGS = Preferences.@load_preference("PERFORMANCE_WARNINGS", true)
+
+"""
+    set_performance_warnings!(b::Bool)
+
+Enable/disable performance warnings globally.
+"""
+function set_performance_warnings!(b::Bool)
+    Preferences.@set_preferences!("PERFORMANCE_WARNINGS" => b)
+    if b
+        @info("Performance Warnings enabled. Restart Julia for this change to take effect.")
+    else
+        @info("Performance Warnings disabled. Restart Julia for this change to take effect.")
+    end
+end
+
+should_warn_paramtype(p::AbstractArray) = !isconcretetype(eltype(p))
+should_warn_paramtype(::Dict{K,V}) where {K,V} = !isconcretetype(V)
+should_warn_paramtype(p::AbstractArray{<:AbstractArray}) = any(should_warn_paramtype, p)
+should_warn_paramtype(p::Tuple) = any(should_warn_paramtype, p)
+should_warn_paramtype(p::NamedTuple) = any(should_warn_paramtype, p)
+should_warn_paramtype(p) = false
+
+const WARN_PARAMTYPE_MESSAGE = """
+Using arrays or dicts to store parameters of different types can hurt performance. \
+Consider using tuples instead.
+"""
+
+"""
+    warn_paramtype(p, warn_performance=PERFORMANCE_WARNINGS)
+
+Inspect the type of `p` and emit a warning if it could hurt
+performance when used to hold problem parameters.
+
+The warning can be turned off by setting `warn_performance` to `false`.
+To turn it off globally, call `set_performance_warnings!(false)`.
```

Comment:
We should tell people to just add a preference to their startup, similar to how it's done in https://docs.sciml.ai/DiffEqDocs/stable/features/low_dep/#Controlling-Function-Specialization-and-Precompilation

Reply:
To clarify, you want me to change the docs to tell people what command to run, and ditch `set_performance_warnings!` ?

Reply:
Yeah, and let them know it can be done in a startup.jl

Reply:
For clarification, I am not sure `set_performance_warnings!` works as suggested, since you'd need to set the preference before the package is `using`ed, but it requires the package has been `using`ed in order to be called. That's why the Preferences system has a way of setting preferences via UUID.